web/pandas/pdeps/0006-ban-upcasting.md

```diff
@@ -0,0 +1,178 @@
+# PDEP-6: Ban upcasting in setitem-like operations
+
+- Created: 23 December 2022
+- Status: Under discussion
+- Discussion: [#50402](https://github.com/pandas-dev/pandas/pull/50402)
+- Author: [Marco Gorelli](https://github.com/MarcoGorelli) ([original issue](https://github.com/pandas-dev/pandas/issues/39584) by [Joris Van den Bossche](https://github.com/jorisvandenbossche))
+- Revision: 1
+
+## Abstract
+
+The suggestion is that setitem-like operations would
+not change a ``Series``' dtype.
+
+Current behaviour:
+```python
+In [1]: ser = pd.Series([1, 2, 3], dtype='int64')
+
+In [2]: ser[2] = 'potage'
+
+In [3]: ser  # dtype changed to 'object'!
+Out[3]:
+0         1
+1         2
+2    potage
+dtype: object
+```
+
+Suggested behaviour:
+
+```python
+In [1]: ser = pd.Series([1, 2, 3])
+
+In [2]: ser[2] = 'potage'  # raises!
+---------------------------------------------------------------------------
+TypeError: Invalid value 'potage' for dtype int64
+```
+
+## Motivation and Scope
+
+Currently, pandas is extremely flexible in handling different dtypes.
+However, this can potentially hide bugs, break user expectations, and copy data
+in what looks like it should be an inplace operation.
+
+An example of it hiding a bug is:
+```python
+In[9]: ser = pd.Series(pd.date_range("2000", periods=3))
+
+In[10]: ser[2] = "2000-01-04"  # works, is converted to datetime64
+
+In[11]: ser[2] = "2000-01-04x"  # typo - but pandas does not error, it upcasts to object
+```
+
+The scope of this PDEP is limited to setitem-like operations on Series.
+For example, starting with
+```python
+df = DataFrame({"a": [1, 2, np.nan], "b": [4, 5, 6]})
+ser = df["a"].copy()
+```
+then the following would all raise:
+- ``ser[0] = 'foo'``;
+- ``ser.fillna('foo', inplace=True)``;
+- ``ser.where(ser.isna(), 'foo', inplace=True)``
+- ``ser.fillna('foo', inplace=False)``;
+- ``ser.where(ser.isna(), 'foo', inplace=False)``
+- ``ser.iloc[0] = 'foo'``
+- ``ser.loc[0] = 'foo'``
+- ``df.loc[indexer, 'a'] = 'foo'``
+- ``ser[indexer] = 'foo'``
+
+where ``indexer`` could be a slice, a mask, a single value, a list or array of values,
+or any other allowed indexer.
+
+Examples of operations which would not raise are:
+- ``ser.diff()``;
+- ``pd.concat([ser, ser.astype(object)])``;
+- ``ser.mean()``;
+- ``ser[0] = 3.``;
+- ``ser[0] = 3``.
+
+## Detailed description
+
+Concretely, the suggestion is:
+- if a ``Series`` is of a given dtype, then a ``setitem``-like operation should not change its dtype;
```

Comment:
For an edge case like

```
ser = Series([1, 2, 3], dtype=np.int64)
ser[0] = np.int8(5)
```

The scalar would be upcast to from int8 to int64 and the result would be int64?

Reply:
sounds reasonable

Reply:
> The scalar would be upcast to from int8 to int64 and the result would be int64?

Just to be explicit: that's also the current behaviour, so that's an example that wouldn't change